File: Sources/Sentry/SentryAutoBreadcrumbTrackingIntegration.m
#import "SentryAutoBreadcrumbTrackingIntegration.h"
#import "SentryBreadcrumbTracker.h"
#import "SentryDependencyContainer.h"
#import "SentryFileManager.h"
#import "SentryLog.h"
#import "SentryOptions.h"
#import "SentrySDK.h"
#import "SentrySystemEventBreadcrumbs.h"

NS_ASSUME_NONNULL_BEGIN

@interface
SentryAutoBreadcrumbTrackingIntegration ()

@property (nonatomic, strong) SentryBreadcrumbTracker *breadcrumbTracker;

#if TARGET_OS_IOS && UIKIT_LINKED
@property (nonatomic, strong) SentrySystemEventBreadcrumbs *systemEventBreadcrumbs;
#endif // TARGET_OS_IOS && UIKIT_LINKED

@end

@implementation SentryAutoBreadcrumbTrackingIntegration

- (BOOL)installWithOptions:(SentryOptions *)options
{
    if (![super installWithOptions:options]) {
        return NO;
    }

    [self installWithOptions:options
             breadcrumbTracker:[[SentryBreadcrumbTracker alloc] init]
#if TARGET_OS_IOS && UIKIT_LINKED
        systemEventBreadcrumbs:
            [[SentrySystemEventBreadcrumbs alloc]
                         initWithFileManager:[SentryDependencyContainer sharedInstance].fileManager
                andNotificationCenterWrapper:[SentryDependencyContainer sharedInstance]
                                                 .notificationCenterWrapper]
#endif // TARGET_OS_IOS && UIKIT_LINKED
    ];

    return YES;
}

- (SentryIntegrationOption)integrationOptions
{
    return kIntegrationOptionEnableAutoBreadcrumbTracking;
}

/**
 * For testing.
 */
- (void)installWithOptions:(nonnull SentryOptions *)options
         breadcrumbTracker:(SentryBreadcrumbTracker *)breadcrumbTracker
#if TARGET_OS_IOS && UIKIT_LINKED
    systemEventBreadcrumbs:(SentrySystemEventBreadcrumbs *)systemEventBreadcrumbs
#endif // TARGET_OS_IOS && UIKIT_LINKED
{
    self.breadcrumbTracker = breadcrumbTracker;
    [self.breadcrumbTracker startWithDelegate:self];

    if (options.enableSwizzling) {
        [self.breadcrumbTracker startSwizzle];
    }

#if TARGET_OS_IOS && UIKIT_LINKED
    self.systemEventBreadcrumbs = systemEventBreadcrumbs;
    [self.systemEventBreadcrumbs startWithDelegate:self];
#endif // TARGET_OS_IOS && UIKIT_LINKED
}

- (void)uninstall
{
    if (nil != self.breadcrumbTracker) {
        [self.breadcrumbTracker stop];
    }
#if TARGET_OS_IOS && UIKIT_LINKED
    if (nil != self.systemEventBreadcrumbs) {
        [self.systemEventBreadcrumbs stop];
    }
#endif // TARGET_OS_IOS && UIKIT_LINKED
}

- (void)addBreadcrumb:(SentryBreadcrumb *)crumb
{
    [SentrySDK addBreadcrumb:crumb];
}

@end

NS_ASSUME_NONNULL_END

File: Sources/Swift/Integrations/AutoBreadcrumbTracking/SentryAutoBreadcrumbTrackingIntegration.swift
@_implementationOnly import _SentryPrivate

typealias AutoBreadcrumbTrackingIntegrationProvider = FileManagerProvider & NotificationCenterProvider

final class SentryAutoBreadcrumbTrackingIntegration<Dependencies: AutoBreadcrumbTrackingIntegrationProvider>: NSObject, SwiftIntegration, SentryBreadcrumbDelegate {
    private let options: Options
    private let fileManager: SentryFileManager
    private let notificationCenterWrapper: SentryNSNotificationCenterWrapper
    private var breadcrumbTracker: SentryBreadcrumbTracker?
    
    #if os(iOS) && !SENTRY_NO_UIKIT
    private var systemEventBreadcrumbs: SentrySystemEventBreadcrumbs?
    #endif // os(iOS) && !SENTRY_NO_UIKIT

    init?(with options: Options, dependencies: Dependencies) {
        guard options.enableAutoBreadcrumbTracking else {
            SentrySDKLog.debug("Not going to enable \(Self.name) because enableAutoBreadcrumbTracking is disabled.")
            return nil
        }

        guard let fileManager = dependencies.fileManager else {
            SentrySDKLog.fatal("File manager is not available")
            return nil
        }

        self.options = options
        self.fileManager = fileManager
        self.notificationCenterWrapper = dependencies.notificationCenterWrapper

        super.init()

        // Create breadcrumb tracker
        let breadcrumbTracker = SentryBreadcrumbTracker(reportAccessibilityIdentifier: options.reportAccessibilityIdentifier)
        self.breadcrumbTracker = breadcrumbTracker
        breadcrumbTracker.start(with: self)

        #if SENTRY_HAS_UIKIT
        if options.enableSwizzling {
            breadcrumbTracker.startSwizzle()
        }
        #endif // SENTRY_HAS_UIKIT

        #if os(iOS) && !SENTRY_NO_UIKIT
        // Create system event breadcrumbs for iOS
        let systemEventBreadcrumbs = SentrySystemEventBreadcrumbs(
            fileManager: fileManager,
            andNotificationCenterWrapper: notificationCenterWrapper
        )
        self.systemEventBreadcrumbs = systemEventBreadcrumbs
        systemEventBreadcrumbs.start(with: self)
        #endif // os(iOS) && !SENTRY_NO_UIKIT
    }

    func uninstall() {
        breadcrumbTracker?.stop()
        breadcrumbTracker = nil
        
        #if os(iOS) && !SENTRY_NO_UIKIT
        systemEventBreadcrumbs?.stop()
        systemEventBreadcrumbs = nil
        #endif // os(iOS) && !SENTRY_NO_UIKIT
    }

    static var name: String {
        "SentryAutoBreadcrumbTrackingIntegration"
    }

    // MARK: - SentryBreadcrumbDelegate

    func add(_ crumb: SentryBreadcrumb) {
        SentrySDKInternal.addBreadcrumb(crumb)
    }
}
